pkg/tso/keyspace_group_manager.go

```diff
@@ -94,63 +130,392 @@ func NewKeyspaceGroupManager(
 	}
 
 	ctx, cancel := context.WithCancel(ctx)
-	ksgMgr := &KeyspaceGroupManager{
+	kgm := &KeyspaceGroupManager{
 		ctx:                         ctx,
 		cancel:                      cancel,
+		tsoServiceID:                tsoServiceID,
 		etcdClient:                  etcdClient,
 		electionNamePrefix:          electionNamePrefix,
-		defaultKsgStorageTSRootPath: defaultKsgStorageTSRootPath,
+		legacySvcRootPath:           legacySvcRootPath,
 		tsoSvcRootPath:              tsoSvcRootPath,
 		cfg:                         cfg,
 		maxResetTSGap:               func() time.Duration { return cfg.GetMaxResetTSGap() },
+		loadKeyspaceGroupsTimeout:   defaultLoadKeyspaceGroupsTimeout,
+		loadKeyspaceGroupsBatchSize: defaultLoadKeyspaceGroupsBatchSize,
 	}
 
-	return ksgMgr
+	kgm.legacySvcStorage = endpoint.NewStorageEndpoint(
+		kv.NewEtcdKVBase(kgm.etcdClient, kgm.legacySvcRootPath), nil)
+	kgm.tsoSvcStorage = endpoint.NewStorageEndpoint(
+		kv.NewEtcdKVBase(kgm.etcdClient, kgm.tsoSvcRootPath), nil)
+	return kgm
 }
 
 // Initialize this KeyspaceGroupManager
-func (kgm *KeyspaceGroupManager) Initialize() {
-	// TODO: dynamically load keyspace group assignment from the persistent storage and add watch for the assignment change
-	kgm.initDefaultKeyspaceGroup()
+func (kgm *KeyspaceGroupManager) Initialize(loadFromStorage bool) error {
+	kgm.legacySvcStorage = endpoint.NewStorageEndpoint(
```

Comment:
Why create the `legacySvcStorage` twice here? It has already been initialized in `NewKeyspaceGroupManager` function.